client/settings/settings.tsx

```diff
@@ -1,228 +1,391 @@
+
+export function ConnectedSettings() {
+  const dispatch = useAppDispatch()
+  const isOpen = useAppSelector(s => s.settings.open)
+  const subPage = useAppSelector(s => s.settings.subPage)
+  const starcraft = useAppSelector(s => s.starcraft)
+
+  const focusableRef = useRef<HTMLSpanElement>(null)
+  const portalRef = useExternalElementRef()
+
+  const onChangeSubPage = useStableCallback((value: SettingsSubPage) => {
+    dispatch(changeSettingsSubPage(value))
+  })
+  const onCloseSettings = useStableCallback(() => {
+    dispatch(closeSettings())
+  })
+
+  const settingsTransition = useTransition<boolean, UseTransitionProps<boolean>>(isOpen, {
+    from: { opacity: 0 },
+    enter: { opacity: 1 },
+    leave: { opacity: 0 },
+    config: {
+      ...defaultSpring,
+      clamp: true,
+    },
+  })
+
+  return ReactDOM.createPortal(
+    settingsTransition((style, isOpen) =>
+      isOpen ? (
+        <KeyListenerBoundary>
+          <FocusTrap focusableRef={focusableRef}>
+            <span ref={focusableRef} tabIndex={-1}>
+              <Settings
+                style={style}
+                subPage={subPage}
+                isStarcraftHealthy={checkIsStarcraftHealthy({ starcraft })}
+                onChangeSubPage={onChangeSubPage}
+                onCloseSettings={onCloseSettings}
+              />
+            </span>
+          </FocusTrap>
+        </KeyListenerBoundary>
+      ) : undefined,
+    ),
+    portalRef.current,
+  )
+}
+
+const Container = styled(animated.div)`
+  position: absolute;
+  top: var(--sb-system-bar-height, 0);
+  left: 0;
+  right: 0;
+  bottom: 0;
+  width: 100%;
+  height: calc(100% - var(--sb-system-bar-height, 0));
+
   display: flex;
-  align-items: center;
-  margin-right: 24px;
+  flex-direction: row;
+
+  background-color: ${background800};
+  z-index: ${zIndexSettings};
 `
 
-const TitleActionText = styled.div`
-  ${body1};
+const NavContainer = styled.div`
+  width: 272px;
+  padding: 16px 0;
+  background-color: ${background700};
+
+  flex-shrink: 0;
+`
+
+const NavSectionTitle = styled.div`
+  ${overline};
+  ${singleLine};
+
+  height: 36px;
+  line-height: 36px;
+  padding: 0 16px;
+
   color: ${colorTextSecondary};
-  margin-right: 4px;
+`
+
+const NavSectionSeparator = styled.div`
+  height: 1px;
+  margin: 7px 16px 8px;
+
+  background-color: ${colorDividers};
+`
+
+function Settings({
+  style,
+  subPage,
+  isStarcraftHealthy,
+  onChangeSubPage,
+  onCloseSettings,
+}: {
+  style: React.CSSProperties
+  subPage: SettingsSubPage
+  isStarcraftHealthy: boolean
+  onChangeSubPage: (subPage: SettingsSubPage) => void
+  onCloseSettings: () => void
+}) {
+  useKeyListener({
+    onKeyDown(event) {
+      if (event.code === ESCAPE) {
+        onCloseSettings()
+        return true
+      }
+
+      return false
+    },
+  })
+
+  return (
+    <Container style={style}>
+      <NavContainer>
+        {IS_ELECTRON ? (
+          <>
+            <NavSectionTitle>App settings</NavSectionTitle>
```

Comment:
Would prefer that everything here (titles, all sections, separators) were just a single data structure that gets mapped over instead of repeating this mapping code a bunch of times. This mapping is very error-prone and it is extremely unlikely someone is going to notice in a review when you mess it up, so I don't want it to exist 4+ times in the code

Reply:
An alternative is to create a function (or a function that creates a function, since we might need to specify that the rendered entries are disabled) that you pass to the `map` call (and fix the one below where you didn't use `map` to render this thing)